apps/manager_app/auth_card.c

```diff
@@ -0,0 +1,542 @@
+
+          /* TODO: Check if card operation failed*/
+          // return MANAGER_TASK_FAILED;
+        }
+
+        UPDATE_AUTH_CARD_FLOW_STATUS(MANAGER_AUTH_CARD_STATUS_PAIRING_DONE);
+        delay_scr_init(ui_text_card_authentication_success, DELAY_TIME);
+
+        return MANAGER_TASK_SUCCESS;
+      }
+      break;
+
+    default:
+      return MANAGER_TASK_INVALID_STATE;
+      break;
+  }
+
+  return MANAGER_TASK_INVALID_DEFAULT;
+}
+
+manager_error_code_t handle_auth_card_query(
+    usb_event_t *usb_evt,
+    manager_auth_card_response_t *resp) {
+  if (NULL == usb_evt || NULL == resp) {
+    return MANAGER_TASK_INVALID_ARGS;
+  }
+
+  manager_query_t query = MANAGER_QUERY_INIT_ZERO;
+
+  if (false ==
+      decode_manager_query(usb_evt->p_msg, usb_evt->msg_size, &query)) {
+    // Ignore invalid data
+    return MANAGER_TASK_DECODING_FAILED;
+  }
+
+  if ((query.which_request != MANAGER_QUERY_AUTH_CARD_TAG)) {
+    // Ignore invalid request
+    return MANAGER_TASK_UNKNOWN_QUERY_REQUEST;
+  }
+
+  switch (query.auth_card.which_request) {
+    case MANAGER_AUTH_CARD_REQUEST_INITIATE_TAG:
+      return handle_auth_card_initiate_query(&(query.auth_card.initiate), resp);
+      break;
+
+    case MANAGER_AUTH_CARD_REQUEST_CHALLENGE_TAG:
+      return handle_auth_card_challenge_query(&(query.auth_card.challenge),
+                                              resp);
+      break;
+
+    case MANAGER_AUTH_CARD_REQUEST_RESULT_TAG:
+      return handle_auth_card_result_query(&(query.auth_card.result));
+      break;
+
+    default:
+      return MANAGER_TASK_UNKNOWN_QUERY_REQUEST;
+      break;
+  }
+
+  return MANAGER_TASK_INVALID_DEFAULT;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+void card_auth_handler(usb_event_t *usb_evt) {
+  ASSERT(NULL != usb_evt);
+
+  evt_status_t evt_status = {.usb_event = *usb_evt};
+  manager_error_code_t query_handler_status = MANAGER_TASK_INVALID_DEFAULT;
+  manager_auth_card_response_t resp = MANAGER_AUTH_CARD_RESPONSE_INIT_ZERO;
+  memzero(&auth_card_ctx, sizeof(auth_card_ctx));
+  UPDATE_AUTH_CARD_FLOW_STATUS(MANAGER_AUTH_CARD_STATUS_INIT);
+
+  do {
+    // Handle auth card query received from USB host
+    query_handler_status =
+        handle_auth_card_query(&(evt_status.usb_event), &resp);
+
+    /* TODO: Create a handler to resolve specific returns and either send error
+     * response or abort flow accordingly*/
+    if (MANAGER_TASK_SUCCESS != query_handler_status &&
+        MANAGER_TASK_UNKNOWN_QUERY_REQUEST != query_handler_status &&
+        MANAGER_TASK_INVALID_STATE != query_handler_status) {
+      return;
+    }
+
+    if (resp.which_response != 0) {
+      send_auth_card_response(&resp);
+      resp.which_response = 0;
+    }
+
+    if (MANAGER_AUTH_CARD_STATUS_PAIRING_DONE == auth_card_ctx.flow_status) {
+      usb_clear_event();
```

Comment:
Hi Aman,

Instead of this special handling, let's propose a change in SDK through which the flow can be concluded by the device. The device will send a dummy message when it has processed the result query.

This will make the communication between device and host consistent (Host query always have a response)


Refer change proposed in device auth flow: https://github.com/Cypherock/cypherock-common/pull/3

Reply:
Will take a look at this, but have to consider that pairing would happen after result query

Reply:
added